### PR TITLE
[NUP] A04 Vsevolod Vaskin

### PR DESCRIPTION
**Pull request by Seva-Vaskin**

### CODE REVIEW

File: src/SM.lama [hunk 1]
```diff
@@ -137,6 +170,27 @@ public fun compileSM (stmt) {
...
+                                          <+> s1compiled <+ JMP(lab) <+> label(label1, true) <+> s2compiled]
+              esac
+          esac
+      esac
+    | While(exp, st)   ->
+  case env.genLabel of
+    [label1, env] ->
+      case compile(label1, env, st) of
+        [s1WasLabel, env, s1compiled] -> [s1WasLabel, env, label(label1, true) <+> compileExpr(exp) 
+                                  <+ CJMP("z", lab) <+> s1compiled <+ JMP(label1) <+> label(lab, true)]
```
Comment (e2e4b6b7) [inline src/SM.lama:186]: label lab have to be created by the function that called compilation of this statement. Responsibility of this statement is to return whether lab was used or not

File: src/SM.lama [hunk 2]
```diff
@@ -137,6 +170,27 @@ public fun compileSM (stmt) {
...
+      esac
+  esac
+    | DoWhile (s1, exp) -> case genLabel(env) of 
+        [label1, env] -> case compile(lab, env, s1) of
```
Comment (e2e4b6b7) [inline src/SM.lama:190]: `compile(lab, env, s1)` if you are compiling body with the lab as a label of end, you saying that the end of body is after the loop, while the end of body is the beginning of the condition check

File: src/SM.lama [hunk 3]
```diff
@@ -137,6 +170,27 @@ public fun compileSM (stmt) {
...
+  case env.genLabel of
+    [label1, env] ->
+      case compile(label1, env, st) of
+        [s1WasLabel, env, s1compiled] -> [s1WasLabel, env, label(label1, true) <+> compileExpr(exp) 
+                                  <+ CJMP("z", lab) <+> s1compiled <+ JMP(label1) <+> label(lab, true)]
+      esac
+  esac
+    | DoWhile (s1, exp) -> case genLabel(env) of 
+        [label1, env] -> case compile(lab, env, s1) of 
+            [s1WasLabel, env, s1compiled] -> [s1WasLabel, env, label(label1, true) <+> s1compiled <+> compileExpr(exp) <+ CJMP("nz", label1)]
```
Comment (e2e4b6b7) [inline src/SM.lama:191]: The same issue with the first return value as for while loop

**Comment by Seva-Vaskin** [comment]
I'm not sure I understood all the comments correctly, but I think I've corrected everything

File: src/SM.lama (hunk 2 above)
Comment (e2e4b6b7) [reply src/SM.lama]: This is not fixed. Now the end of the body is the beginning of the body (you skipped check of the loop condition). You need one more label between body and condition

**Comment by e2e4b6b7** [comment]
the rest is good

**Comment by Seva-Vaskin** [comment]
DoWhile is fixed

**Review by e2e4b6b7 — APPROVED** [review]

**Pull request closed without merging**